streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -2316,12 +2317,14 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
 
         lazy val transducer: ZChannel[R1, E, Chunk[A1], Any, E1, Chunk[Z], Unit] =
           sink.channel.doneCollect.flatMap { case (leftover, z) =>
-            ZChannel.fromEffect(leftovers.set(leftover.flatten)) *>
-              ZChannel.write(Chunk.single(z)) *>
-              ZChannel.fromEffect(upstreamDone.get).flatMap { done =>
-                if (done) ZChannel.end(())
-                else transducer
-              }
+            ZChannel.fromEffect(upstreamDone.get <*> leftovers.updateAndGet(_ ++ leftover.flatten)).flatMap {
+              case (done, newLeftovers) =>
+                val nextChannel =
+                  if (done && newLeftovers.isEmpty) ZChannel.end(())
```

Comment:
There's a potential non-termination condition here. If a sink keeps returning leftovers, transduce will never terminate. Interested in hearing other opinions. The options I thought of are:
* This behavior
* If the stream is done, feed the leftovers to the sink one more time and check if the returned leftovers are identical, and if so, terminate (while dropping the leftovers)

Reply:
I think if the sink returns the leftovers, it is a sign that the sink needs more to continue. So feeding them to the sink again should not change the result. In fact that almost seems like a law of a well-behaved sink: if it returns leftovers, and then is fed those leftovers again, it must again return them.